File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char

isShouting :: String -> Bool
isShouting m = hasAlpha && all isUpper alphaOnly
  where alphaOnly = filter isAlpha m
        hasAlpha  = not (null alphaOnly)

isQuestion :: String -> Bool
isQuestion m = last m == '?'

isSilence :: String -> Bool
isSilence m  = all isSpace m

responseFor :: String -> String
responseFor message
  | isSilence  message = "Fine. Be that way!"
  | isShouting message = "Woah, chill out!"
  | isQuestion message = "Sure."
  | otherwise          = "Whatever."

File: elixir/bob/bob.exs
defmodule Teenager do
  @moduledoc"""
  Responds to inputs in a surly way
  """
  @doc"""
  Responds to inputs in a surly way
  """
  def hey(input) do
    cond do
      String.length(String.rstrip(input)) == 0 ->
        "Fine. Be that way!"
      String.upcase(input) == input && String.contains?(input, all_the_letters) ->
        "Woah, chill out!"
      String.last(input) == "?" ->
        "Sure."
      true ->
        "Whatever."
    end
  end

  defp all_the_letters do
    String.codepoints(to_string(Enum.to_list(?A..?Z) ++ Enum.to_list(?a..?z)))
  end
end
